### PR TITLE
Unified human FG name files

### PR DESCRIPTION
**Pull request by P-D-E**
Thank you for submitting a pull request and becoming a contributor to Vega Strike: Upon the Coldest Sea.

Please answer the following:

Code Changes:
- [x] Have the PR Validation Tests been run? See https://github.com/vegastrike/Vega-Strike-Engine-Source/wiki/Pull-Request-Validation
- [ ] This is a documentation change only

Issues:
- 

Purpose:
- What is this pull request trying to do?  **Replace duplicate flight group file names with a new standard one for all the human factions**
- What release is this for?  **0.8.x**
- Is there a project or milestone we should apply this to?  **0.8.x**

### CODE REVIEW

**Review by Loki1950 — APPROVED** [review]

**Comment by BenjamenMeyer** [comment]
@P-D-E looks good.

File: modules/fg_util.py
Comment (P-D-E) [reply modules/fg_util.py:76]: If you're okay with this,
```    names = [filename,
             os.path.join('../', filename),
             os.path.join('universe', 'fgnames', 'names.txt'),
             os.path.join('..', 'universe', 'names.txt'),
             os.path.join('universe', 'names.txt)'
             ]
```
I can commit it right now.

Comment (stephengtuggy) [reply modules/fg_util.py:76]: @P-D-E that looks good to me. I would go ahead and commit it.

Sorry I took so long to get back to you on this PR. With this change, I think I will go ahead and approve it.

Comment (P-D-E) [reply modules/fg_util.py]: Done, I also changed the `    filename = 'universe/fgnames/'+faction+'.txt'` line above it to use `os.path.join` accordingly.

**Review by stephengtuggy — APPROVED** [review]
LGTM!

**Comment by stephengtuggy** [comment]
Has this been play tested yet?

**Comment by P-D-E** [comment]
@stephengtuggy I didn't follow the standard playtest and did this:

- take away the human factions' files from `universe/fgnames` leaving only `names.txt`
- add a print line to report faction and filename with relative path here
 ```
     for name in names:
         try:
             with open (name,'r') as f:
                 bnl = f.readlines()
             print(faction, 'names found as', name, 'e.g.', bnl[0])  # <--- this one
             break
 ```
- play a game with at least a few human factions in the sector, checking the terminal for that print at every occurence; `names.txt` has been used for the factions without a corresponding file.

**Comment by stephengtuggy** [comment]
@P-D-E So you did some testing specific to this change set. Great.

I think we will still want to run the standard set of tests from https://github.com/vegastrike/Vega-Strike-Engine-Source/wiki/Pull-Request-Validation before merging this. Also, did you check to make sure that, e.g. Rlaan factions still work properly?

**Comment by P-D-E** [comment]
> @P-D-E So you did some testing specific to this change set. Great.
> did you check to make sure that, e.g. Rlaan factions still work properly?

Of course, that was the point of the test; `names.txt` has been used for the factions without a corresponding file, not for every faction in game.

**Review by BenjamenMeyer — APPROVED** [review]
LGTM pending a standard play test

**Comment by stephengtuggy** [comment]
@P-D-E Yes, I think we'll need to address that first issue with a fallback mechanism, before this PR gets merged.

As for the second issue, I've already filed a ticket for that here: https://github.com/vegastrike/Vega-Strike-Engine-Source/pull/537

**Comment by P-D-E** [comment]
> @P-D-E Yes, I think we'll need to address that first issue with a fallback mechanism, before this PR gets merged.

This does the job, is it acceptable?
```
    base_faction = faction.replace('_citizen', '').replace('_guild', '').replace('_briin', '').replace('n_merchant_marine', '')
    base_filename = os.path.join('universe', 'fgnames', base_faction + '.txt')
    names = [
             filename,
             base_filename,
             os.path.join('../', filename),
             os.path.join('../', base_filename),
             os.path.join('universe', 'fgnames', 'names.txt'),
             os.path.join('..', 'universe', 'names.txt'),
             os.path.join('universe', 'names.txt')
            ]
```

> As for the second issue, I've already filed a ticket for that here: [vegastrike/Assets-Masters#30](https://github.com/vegastrike/Assets-Masters/issues/30)

Good!

**Comment by stephengtuggy** [comment]
> > @P-D-E Yes, I think we'll need to address that first issue with a fallback mechanism, before this PR gets merged.
> 
> This does the job, is it acceptable?
> 
> ```
>     base_faction = faction.replace('_citizen', '').replace('_guild', '').replace('_briin', '').replace('n_merchant_marine', '')
>     base_filename = os.path.join('universe', 'fgnames', base_faction + '.txt')
>     names = [
>              filename,
>              base_filename,
>              os.path.join('../', filename),
>              os.path.join('../', base_filename),
>              os.path.join('universe', 'fgnames', 'names.txt'),
>              os.path.join('..', 'universe', 'names.txt'),
>              os.path.join('universe', 'names.txt')
>             ]
> ```

Looks good to me. @Loki1950 , can you take a look at the above code snippet?

**Comment by Loki1950** [comment]
Looks good to me.

**Comment by P-D-E** [comment]
For the record, here are a few debugging prints showing the fallback in action:
```
merchant_guild names found as universe/fgnames/merchant.txt e.g. Aachen
rlaan_briin names found as universe/fgnames/rlaan.txt e.g. Abelia_x_grandiflora
aeran_merchant_marine names found as universe/fgnames/aera.txt e.g. abelsonite
rlaan_citizen names found as universe/fgnames/rlaan.txt e.g. Abelia_x_grandiflora
merchant_guild_citizen names found as universe/fgnames/merchant.txt e.g. Aachen
```

**Comment by stephengtuggy** [comment]
@P-D-E if it's good enough for @Loki1950 , it's good enough for me. Why don't you go ahead and commit that change, and then I think we'll be ready to merge this PR.

**Comment by stephengtuggy** [comment]
(I was mainly concerned about whether the `rlaan` and `rlaan_briin` should use the same set of faction names or not)

**Comment by stephengtuggy** [comment]
On second thought: I'm getting a crash on Windows. Let me investigate a little further. We might have an OS-specific problem here.

**Comment by stephengtuggy** [comment]
> On second thought: I'm getting a crash on Windows. Let me investigate a little further. We might have an OS-specific problem here.

I'm getting the following on my Windows machine:

```
Loading active missions True
void Mission::DirectorLoop(): Python error occurred
Traceback (most recent call last):
  File "C:\devel\Assets-Production\modules\missions\privateer.py", line 34, in Execute
    i.Execute()
  File "C:\devel\Assets-Production\modules\random_encounters.py", line 290, in Execute
    generate_dyn_universe.KeepUniverseGenerated()
  File "C:\devel\Assets-Production\modules\generate_dyn_universe.py", line 312, in KeepUniverseGenerated
    ReloadUniverse()
  File "C:\devel\Assets-Production\modules\generate_dyn_universe.py", line 277, in ReloadUniverse
    GenerateAllShips() ###Insert number of flight groups and max ships per fg
  File "C:\devel\Assets-Production\modules\generate_dyn_universe.py", line 80, in GenerateAllShips
    fgnames.append(fg_util.GetRandomFGNames(-1,faction_ships.factions[fnr]))
  File "C:\devel\Assets-Production\modules\fg_util.py", line 101, in GetRandomFGNames
    flightgroupnamelist[faction]=ReadBaseNameList(faction)
  File "C:\devel\Assets-Production\modules\fg_util.py", line 74, in ReadBaseNameList
    filename = os.path.join('universe', 'fgnames', faction + '.txt')
NameError: name 'os' is not defined
```

Do we maybe just need an `import os` line near the top?

**Comment by P-D-E** [comment]
> (I was mainly concerned about whether the `rlaan` and `rlaan_briin` should use the same set of faction names or not)

The '_briin' substitution can be taken away as soon as the rlaan_briin.txt is provided; meanwhile I think it makes more sense to fall back to the main species file.

> Do we maybe just need an `import os` line near the top?

Oops, forgot to commit that one! Done.

**Comment by stephengtuggy** [comment]
> > > @P-D-E Yes, I think we'll need to address that first issue with a fallback mechanism, before this PR gets merged.
> > 
> > 
> > This does the job, is it acceptable?
> > ```
> >     base_faction = faction.replace('_citizen', '').replace('_guild', '').replace('_briin', '').replace('n_merchant_marine', '')
> >     base_filename = os.path.join('universe', 'fgnames', base_faction + '.txt')
> >     names = [
> >              filename,
> >              base_filename,
> >              os.path.join('../', filename),
> >              os.path.join('../', base_filename),
> >              os.path.join('universe', 'fgnames', 'names.txt'),
> >              os.path.join('..', 'universe', 'names.txt'),
> >              os.path.join('universe', 'names.txt')
> >             ]
> > ```
> 
> Looks good to me. @Loki1950 , can you take a look at the above code snippet?

@P-D-E Once you commit the above change, I think this will be ready to merge.

**Comment by P-D-E** [comment]
> @P-D-E Once you commit the above change, I think this will be ready to merge. Pending a brief play test or two.
@stephengtuggy Standard play test passed again, on Linux (https://github.com/vegastrike/Assets-Production/pull/74 included); feel free to test on the other platforms.

**Comment by stephengtuggy** [comment]
@P-D-E Cool.

The only thing is, I don't see the following line in this PR yet:

```py
base_faction = faction.replace('_citizen', '').replace('_guild', '').replace('_briin', '').replace('n_merchant_marine', '')
```

**Comment by P-D-E** [comment]
@stephengtuggy Done

**Review by stephengtuggy — APPROVED** [review]
Cool! There we go.

File: modules/fg_util.py
Comment (stephengtuggy) [inline modules/fg_util.py:89]: Looks like on Windows, you have to specify a code page, or else Python 3 will default to cp1252 for some reason.

Try replacing this line with:

```py
with open (name, 'r', encoding='utf8') as f:
```

Comment (P-D-E) [reply modules/fg_util.py]: @stephengtuggy Done, of course it keeps working on Linux as well.

**Pull request merged**
